File: src/server/util/ServerLogStream.php
<?php

namespace pocketcloud\cloud\server\util;

use pocketcloud\cloud\server\CloudServer;
use RuntimeException;

final class ServerLogStream {

    private bool $startedStream = false;
    private mixed $fileHandle = null;

    public function __construct(private readonly CloudServer $server) {}

    public function startStream(): void {
        if (!@file_exists($this->server->getLogFilePath())) throw new RuntimeException("Log file does not exist");
        $fileHandle = fopen($this->server->getLogFilePath(), "r");
        if ($fileHandle === false) throw new RuntimeException("Log file cannot be opened");
        $this->fileHandle = $fileHandle;
        $this->startedStream = true;
    }

    public function readNewLine(): string|false|null {
        if (!$this->startedStream) return false;
        $line = fgets($this->fileHandle);
        if ($line === false) {
            clearstatcache(false, $this->server->getLogFilePath());
            $currentPos = ftell($this->fileHandle);
            fseek($this->fileHandle, $currentPos);
            $line = fgets($this->fileHandle);
            if ($line === false) return null;
        }

        return trim($line);
    }

    public function stopStream(): void {
        if ($this->startedStream && is_resource($this->fileHandle)) {
            fclose($this->fileHandle);
        }

        $this->startedStream = false;
        $this->fileHandle = null;
    }
}

File: src/main/java/de/pocketcloud/cloud/server/util/ServerLogStream.java
package de.pocketcloud.cloud.server.util;

import de.pocketcloud.cloud.server.CloudServer;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;

public final class ServerLogStream {

    private final CloudServer server;
    private RandomAccessFile fileHandle;
    private boolean startedStream = false;

    public ServerLogStream(CloudServer server) {
        this.server = server;
    }

    public void startStream() {
        File logFile = server.logFilePath().toFile();
        if (!logFile.exists()) throw new RuntimeException("Log file does not exist");
        try {
            this.fileHandle = new RandomAccessFile(logFile, "r");
            this.startedStream = true;
        } catch (FileNotFoundException e) {
            throw new RuntimeException("Log file cannot be opened", e);
        }
    }

    public String readNewLine() {
        if (!startedStream) return null;
        try {
            String line = fileHandle.readLine();
            if (line == null) {
                long currentPos = fileHandle.getFilePointer();
                fileHandle.seek(currentPos);
                line = fileHandle.readLine();
                if (line == null) return null;
            }

            return line.trim();
        } catch (IOException e) {
            return null;
        }
    }

    public void stopStream() {
        if (startedStream && fileHandle != null) {
            try {
                fileHandle.close();
            } catch (IOException _) {}
        }

        startedStream = false;
        fileHandle = null;
    }
}